### ISSUE TITLE
[Feature Request] Model bring up for Gemma2 (TtGemmaAttention)

### ISSUE THREAD

**Issue by ashyuon-cw** [tenstorrent/tt-metal#18919]
In Gemma2 Attention, there are several parameters that do not exist in Llama3, namely: `query_pre_attn_scalar`, `attn_logit_softcapping`, and `sliding_window_size`(slidng_window in HF Config).

For `query_pre_attn_scalar`, I believe it can be incorporated by modifying the `self.scale` part in the `__init__` function of `llama_attention.py` as follows:
```
if configuration.query_pre_attn_scalar is not None:
    self.scale = configuration.query_pre_attn_scalar**-0.5
else:
    self.scale = self.head_dim**-0.5
```

`attn_logit_softcapping` and `sliding_window_size` are used during the attention computation. From my understanding, the functions `ttnn.transformer.paged_scaled_dot_product_attention_decode`, `ttnn.transformer.scaled_dot_product_attention_decode`, `ttnn.transformer.chunked_scaled_dot_product_attention`, and `ttnn.transformer.scaled_dot_product_attention` in `llama_attention.py` handle the entire attention operation, including computing the dot product of query and key, applying scaling, obtaining scores through softmax, and performing a weighted sum with value.

Therefore, to incorporate these parameters, modifications to the corresponding `ttnn` functions would be required. It seems necessary to include the following functionality after performing the dot product of query and key in the ttnn attention function:

```
# Reference: https://github.com/google/gemma_pytorch/blob/main/gemma/model.py#L308
all_ones = torch.ones_like(mask)
sliding_mask = torch.triu(
    all_ones, -1 * self.sliding_window_size + 1
) * torch.tril(all_ones, self.sliding_window_size - 1)
mask = torch.where(sliding_mask == 1, mask, -2.3819763e38)

if self.attn_logit_softcapping is not None:
    scores = scores / self.attn_logit_softcapping
    scores = torch.tanh(scores)
    scores = scores * self.attn_logit_softcapping
if self.attn_logit_softcapping is not None:
    scores = scores / self.attn_logit_softcapping
    scores = torch.tanh(scores)
    scores = scores * self.attn_logit_softcapping
```

Would it be possible to add this functionality? If it has already been implemented, could you let me know how to use it?

Thank you.

**Comment by milank94** [comment]
@uaydonat @rjakovljevicTT related issue on Gemma2 model bring-up

**Comment by ashyuon-cw** [comment]
It seems that there are additional changes in the Gemma3 architecture. This refers to the changes in how LayerNorm is applied to Q and K, as well as how the sliding window mask is handled.
```
# https://github.com/google/gemma_pytorch/blob/main/gemma/model.py#L285
if self.query_norm is not None and self.key_norm is not None:
            xq = self.query_norm(xq)
            xk = self.key_norm(xk)

...

# https://github.com/google/gemma_pytorch/blob/main/gemma/model.py#L317
if (
            self.attn_type == gemma_config.AttentionType.LOCAL_SLIDING
            and self.sliding_window_size is not None
            and local_mask is not None
        ):
            mask = local_mask
```

**Comment by ashyuon-cw** [comment]
The Local and Global layers in Gemma3 follow a repeating structure with a 5:1 ratio.
When implementing the Gemma model, this should be taken into account when configuring the Decoder layers that include the Attention layers.

```
# https://github.com/google/gemma_pytorch/blob/main/gemma/model.py#L672

for i in range(config.num_hidden_layers):
            if config.architecture == gemma_config.Architecture.GEMMA_1:
                self.layers.append(GemmaDecoderLayer(config))
            elif config.architecture in (
                gemma_config.Architecture.GEMMA_2,
                gemma_config.Architecture.GEMMA_3,
            ):
                attn_type = (
                    config.attn_types[i % len(config.attn_types)]
                    if config.attn_types is not None
                    else gemma_config.AttentionType.GLOBAL
                )
                self.layers.append(Gemma2DecoderLayer(config, attn_type))
```

**Comment by ppetrovicTT** [comment]
Gemma3 is brought up and available in:
- Vision: https://github.com/tenstorrent/tt-metal/tree/main/models/demos/gemma3/
- Language: https://github.com/tenstorrent/tt-metal/tree/main/models/tt_transformers/
